tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java

```diff
@@ -54,6 +56,15 @@
     private final DataInputStream dataStream;
     private final ExecutorService executor;
 
+    enum State {
+        Opened,
+        Closed
+    }
+
+    private AtomicReferenceFieldUpdater<BlobStoreBackedReadHandleImpl, State> STATE_UPDATER = AtomicReferenceFieldUpdater
+        .newUpdater(BlobStoreBackedReadHandleImpl.class, State.class, "state");
+    private volatile State state = null;
```

Comment:
I'm not sure that we need these concurrency controls. The `executor` in this class is a single thread. Therefore, a local `state` variable should be sufficient as long as we only update the variable from within the `executor`'s thread.

Reply:
Yes. We did don't need the `AtomicReferenceFieldUpdater` to control it. But we need to make it as `volatile`, the read and close in the different thread, if the read thread into an infinite loop, then someone triggers the close, the read thread can not exit the infinite loop because the state is holding by the read thread. We need to add the volatile to make the state can be notified when it's changed.

Reply:
> the read and close in the different thread, if the read thread into an infinite loop, then someone triggers the close, the read thread can not exit the infinite loop because the state is holding by the read thread.

The read and the write are not in separate threads. They are both scheduled to run as runnables on the `executor` in this class. That `executor` has a single thread. If the read is in an infinite loop, the `close` runnable will _never_ run.